File: Haskell/data_structures/11-ValidBinarySearchTree.hs
import Control.Monad (replicateM)
import Data.Vector (Vector(..), (!), fromList)

data Node = Node { key :: Int, leftIx :: Int, rightIx :: Int }
data Tree = Tree { nodes :: Vector Node, rootIx :: Int }


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readNodeInformation :: Int -> IO (Vector Node)
readNodeInformation nrNodes = do
    lines' <- replicateM nrNodes getLine
    return $ fromList $ map ((\[key, left, right] -> Node key left right) . convertToIntList) lines'


inorderTraversal :: Tree -> [Int]
inorderTraversal (Tree nodes rootIx) = go [] rootIx []
    where
        go :: [Int] -> Int -> [Node] -> [Int]
        go keys nodeIx stack
            | nodeIx /= -1 = let node = nodes ! nodeIx in go keys (leftIx node) (node:stack)
            | otherwise = case stack of [] -> reverse keys
                                        (Node key _ rightIx):rest -> go (key:keys) rightIx rest


isSorted :: [Int] -> Bool
isSorted [] = True
isSorted [_] = True
isSorted (a:b:rest) = if a <= b then isSorted (b:rest) else False


isValidBinarySearchTree :: Tree -> Int -> Bool
isValidBinarySearchTree tree nrNodes
    | nrNodes <= 1 = True
    | otherwise = let inorderKeys = inorderTraversal tree in isSorted inorderKeys


main :: IO()
main = do
    nrNodes <- readLn
    nodes <-readNodeInformation nrNodes
    let tree = Tree nodes 0
    let verdict = isValidBinarySearchTree tree nrNodes
    print $ if verdict then "CORRECT" else "INCORRECT"

File: Rust/data_structures/src/bin/11_valid_binary_search_tree.rs
use std::io;

struct Node {
    key: i32,
    left_ix: i32,
    right_ix: i32,
}

struct BinaryTree {
    root_ix: i32,
    nodes: Vec<Node>,
}

impl BinaryTree {
    fn new(nodes: Vec<Node>) -> Self {
        BinaryTree { root_ix: 0, nodes }
    }

    fn inorder_traversal(&self) -> Vec<i32> {
        let mut keys = Vec::new();
        let mut stack = Vec::new();
        let mut node_ix = self.root_ix;
        loop {
            if node_ix != -1 {
                let node = &self.nodes[node_ix as usize];
                stack.push(node);
                node_ix = node.left_ix;
            } else {
                let Some(node) = stack.pop() else {
                    return keys
                };
                keys.push(node.key);
                node_ix = node.right_ix;
            }
        }
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line.split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    line
}

fn read_nodes(nr_nodes: usize) -> Vec<Node> {
    let mut nodes = Vec::with_capacity(nr_nodes);
    for _ in 0..nr_nodes {
        let line = read_line();
        let xs: Vec<i32> = convert_to_int_vector(&line);
        let [k, l, r] = xs.as_slice() else {
            panic!("Malformed input: {}", line);
        };
        nodes.push(Node { key: *k, left_ix: *l, right_ix: *r });
    }

    nodes
}

fn is_valid_binary_search_tree(tree: BinaryTree, nr_nodes: usize) -> bool {
    if nr_nodes <= 1 {
        return true
    }

    let keys = tree.inorder_traversal();
    keys.is_sorted()
}

fn main() {
    let nr_nodes = read_line().trim().parse().unwrap();
    let nodes = read_nodes(nr_nodes);
    let tree = BinaryTree::new(nodes);
    let result = is_valid_binary_search_tree(tree, nr_nodes);
    if result {
        println!("CORRECT");
    } else {
        println!("INCORRECT");
    }
}
